python/ray/dataframe/dataframe.py

```diff
@@ -112,6 +118,10 @@ def __init__(self, data=None, index=None, columns=None, dtype=None,
                 elif col_partitions is not None:
                     axis = 1
                     partitions = col_partitions
+                    # All partitions will already have correct dtypes
+                    self._dtypes_cache = [
```

Comment:
Is it more performant to have a special case here for `col_partitions`? Otherwise I'd imagine this is already handled by the added code below.

Reply:
I think so. Otherwise, we create block partitions from column partitions, materialize column partitions again to correct dtypes, and split that back into block partitions.